### ISSUE TITLE
Round-robin proxy check scheduling is inconsistent around restarts

### ISSUE THREAD

**Issue by echlebek** [sensu/sensu-go#4013]
## Expected Behavior
Agent restarts are minimally disruptive to proxy check scheduling.

## Current Behavior
Agent restarts can result in proxy check scheduling being unstable for a large amount of time. In one user's case, it took up to 30 minutes for the schedule to stabilize after a rolling restart of the agents.

## Possible Solution
Unknown.

## Steps to Reproduce (for bugs)
1. Create a cluster with several hundred agents.
2. Schedule round robin proxy checks to run on the agents.
3. Observe the schedule being inconsistent.

## Context
Reported by a customer.

**Comment by portertech** [comment]
We need to do some deep investigative work here. It's a complex issue.

**Comment by portertech** [comment]
I'll schedule a discussion later this week for the team to discuss this issue.

**Comment by mcbsd** [comment]
Hi,

just to share that we are facing this issue for very long time. Below is screenshot of 3 sensu backend nodes where one of them is always over utilized compare to other two where load is distributed more evenly. Some time when the load was too high on single node, we stopped the agent who schedules checks on behalf of proxy entities, then started after a while and the load was distributed more evenly. The sharp drop on the screenshot represents reload of checks where we restart sensu agents and our polling backend on server nodes as part of the reload.
We have 340 proxy entities and 70 proxy checks with splay 90. 

![image](https://user-images.githubusercontent.com/64481947/97678635-6e1fec80-1a94-11eb-868f-1db0074ac859.png)

**Comment by jsnod** [comment]
Just experienced this last night and spent hours trying to figure out why checks weren't triggering.  Proxy checks would all fail with the following:

```json
{
    "component": "schedulerd",
    "level": "error",
    "msg": "check not published, no matching entities for proxy request",
    "name": "check-ecs-service-health",
    "namespace": "default",
    "scheduler_type": "round-robin interval",
    "time": "2020-12-15T23:37:30Z"
}
```

I was able to work around this by setting `round_robin: false` on the check. With `round_robin` off they all started working.

**Comment by echlebek** [comment]
@jsnod this is a different problem from the issue. You're encountering expected behaviour, because you asked for round robin scheduling, but your check didn't match any proxy entities for publishing.

https://docs.sensu.io/sensu-go/latest/observability-pipeline/observe-schedule/checks/#round-robin-checks

**Comment by jsnod** [comment]
@echlebek Perhaps, but I experienced this same issue described by OP until I disabled round_robin. Maybe it's a red herring, but in what situation would a check not match any proxy entities? I have a very simple setup: 1 agent, a couple of checks subscribed to it, and a couple of proxy entities. I'm expecting every check to execute against every one of my entities on my single agent. After restarting my agent, a handful of checks would not execute and I saw a bunch of the above errors.  After some time (30ish min?) they'd all eventually work as described by the OP. But with round_robin off they trigger immediately after agent restart. Might be coincidence but they seem related to me.

**Comment by echlebek** [comment]
@jsnod I see, perhaps it is related after all then. Can you share your check and entity configuration, for reference?

**Comment by jsnod** [comment]
@echlebek Entity/Check config is below.  Also a question I had, does `round_robin` work if you only have 1 agent?

Entity:

```
---
type: Entity
api_version: core/v2
metadata:
  name: my-service
  namespace: default
  labels:
    proxy_type: ecs
    environment: prod
    service_name: my-service
    health_url: https://my-service.com/health
spec:
  entity_class: proxy
```

Check:

```
---
type: CheckConfig
api_version: core/v2
metadata:
  name: check-alb-target-group-health-prod
  namespace: default
spec:
  command: check-alb-target-group-health.rb -t {{ .labels.environment }}-{{ .labels.service_name }}
  handlers:
  - email
  - slack
  interval: 120
  proxy_requests:
    entity_attributes:
    - entity.entity_class == 'proxy'
    - entity.labels.proxy_type == 'ecs'
    - entity.labels.environment == 'prod'
  publish: true
  round_robin: false
  runtime_assets:
  - sensu-plugins-aws
  - sensu-ruby-runtime
  subscriptions:
  - proxy_prod
```

**Comment by echlebek** [comment]
@jsnod it is supposed to, however, there were some reports about round robin scheduling not working correctly with a single agent previously. However, I think the issue was closed out as we couldn't manage to reproduce it. https://github.com/sensu/sensu-go/issues/2926

Perhaps there really is an issue, and you're running into it.

**Comment by portertech** [comment]
Hello everyone :wave: Thank you for your input on this issue. The Sensu Engineering team now has an understanding of why round-robin scheduling is inconsistent and potentially unreliable in OSS builds of Sensu Go. @echlebek has written up an excellent summary of our understanding and provides some advice:

Experience has shown that in some circumstances, sensu-go's scheduling of round robin and proxy round robin checks can be unreliable after cluster or agent restarts. Check scheduling can be delayed or incorrect. Detailed investigation has revealed that the use of the etcd "lease" mechanism for this type of scheduling can lead to this unreliable scheduling.

When sensu-go uses round-robin scheduling, it creates sets of agent entities called "rings" to execute checks on. These rings are composed of one or more leased keys in etcd. The intention is that unless the key is continually updated (i.e. on its keepalive schedule) the key will be removed from the ring, and the round robin scheduler will not consider it for execution. Thus, whenever an agent severs its connection to the server, its ring key is immediately removed. If the connection appears to be maintained, but the agent is unresponsive, then its key will be removed after the lease expires.

In a stable cluster, this method generally works. However, there are scenarios where this method breaks down and stops working.

1. etcd cluster is shut down, or cluster loses majority. In this circumstance, leases can be lost by etcd. Depending on the version of etcd, and depending on the length of the leased keys, the lease may be persisted to storage, or it may be held in memory. If it is held in memory, then the lease will be lost when the cluster restarts. When this happens, these stale ring keys can lead to check requests being sent into the void, with no corresponding agent to receive them, if the agents corresponding to those keys are no longer available.
2. Agents are mass-disconnected from a cluster that is on the brink of failure. If etcd is experiencing performance problems, and a large number of agents disconnect suddenly, then the cluster can fail to deal with the incoming load, and fail to remove all of the leased keys. This usually leads to a cluster restart as well, and then the above situation applies.
3. etcd is under heavy sustained load. When etcd is under intense, sustained load, then lease expirations are given a lower priority than serving GET and PUT requests. This can lead to the system expiring leases only after a very long time passes, sometimes 30-60 minutes. If a large number of agents have been removed from a subscription, this can have disastrous effects on round robin scheduling.

Because of the asynchronous nature of Sensu, it's difficult to detect and deal with this problem. Sensu doesn't associate round robin scheduling requests with check results in any way. So after sending a check request to a subscription, Sensu will never know if any result was delivered for it.

In the commercial distribution of Sensu, we have developed a new round robin scheduling mechanism on top of postgresql, which isn't supported by the OSS sensu-go. We expect this scheduling mechanism to be reliable across sensu-backend and postgres restarts, as there are no in-memory portions of the design.

There may be a way to solve the issue for sensu-go as well, by correlating incoming events with their scheduler, but this would involve a ground-up redesign of the sensu-go scheduler and we aren't ready to tackle that at this time.

Users that need to rely on sensu-go's round robin check scheduling can mitigate the problem somewhat by following the advice below:

- Use only high performance storage for etcd (NVMe SSDs).
- Run etcd separately from sensu-backend, in order to maximize its uptime.
- Run a single etcd node (do not cluster).
- Ensure that the steady-state CPU usage of etcd is low enough to handle bursts of traffic.
- Keep the number of agents per subscriptions relatively small.
- Don't use ephemeral agents (deregister: true). This can lead to more ring removals than with persistent agents (deregister: false), and destabilize round-robin scheduling.

**Comment by portertech** [comment]
At this time, we have a fix that requires the commercial postgres store. Users using the commercial distribution of consistent and reliable round-robin scheduling. As @echlebek mentioned in his summary, there may be a way to solve the issue in sensu-go, we welcome ideas and contributions. Closing this issue for now.
